### ISSUE TITLE
Different json key and PHP variable naming

### ISSUE THREAD

**Issue by o15a3d4l11s2** [JsonMapper/JsonMapper#74]
I have a json structure that has a property called `categories[]`. The problem is that this thing cannot be mapped, because it contains special characters in the name, so I cannot declare `$categories[]` as variable name in PHP.

Having a possibility to add manual mapping description would be very useful. I can imagine this being used in many various scenarios where developers do not want the json key and the PHP variable having the exact same name.

**Comment by DannyvdSluijs** [comment]
That sounds like a good idea. Any change you can share the JSON structure and PHP model to see what you’re trying to achieve?

Would you be making a PR or need help? Or would you for me to take a look?

**Comment by o15a3d4l11s2** [comment]
Unfortunately at the moment I would not be able to submit a PR for this.

Here is an example json (simplified):
```json
{
   "products_highlight":{
      "type":"categories",
      "ids[]":[
         1,
         2,
         3
      ]
   }
}
```

```php
class ProductsHighlight
{
    public ?string $type;
    public ?array $ids;
}
```

Now this cannot be mapped. I know this looks like an edge case, but imagine something like
```json
{
   "products_highlight":{
      "highlight_type":"categories"
   }
}
```
and you want to map `highlight_type ` to a variable named `$type`. That is not possible also at the moment. Please correct me if I am wrong. The only key <-> variable name modification that can be done is middleware for the case conversion.

**Comment by DannyvdSluijs** [comment]
Seems like you’re using PHP 8?? Perhaps the MapFrom middleware (uses PHP 8 and the new attributes) is what you are looking for. It is undocumented in the website but you can see a working example in the integration tests. See https://github.com/JsonMapper/JsonMapper/blob/develop/tests/Integration/Middleware/Attributes/AttributeTest.php

**Comment by o15a3d4l11s2** [comment]
At the moment I am using PHP 7.4, which introduced typed properties.
What you sent is awesome, it is exactly what I was proposing. I did not know it existed already, sorry.
The only downside (for me personally) is that is is for PHP 8 only. Is there a way to define my custom middleware that can do the "map from" work for me prior to PHP 8?

**Comment by DannyvdSluijs** [comment]
I've just released version 2.2.0 which contains a new middleware called `Rename` you can find an example in the [integration test](https://github.com/JsonMapper/JsonMapper/blob/develop/tests/Integration/JsonMapperTest.php#L463) on how to use it.

**Comment by o15a3d4l11s2** [comment]
Thank you @DannyvdSluijs, I really appreciate this. I hope that other users will benefit from this change too.

**Comment by DannyvdSluijs** [comment]
You're welcome @o15a3d4l11s2 Could you help JsonMapper by giving it a star here on GitHub and follow us on [Twitter](http://twitter.com/jsonmapper) perhaps?

**Comment by wntechs** [comment]
> I've just released version 2.2.0 which contains a new middleware called `Rename` you can find an example in the [integration test](https://github.com/JsonMapper/JsonMapper/blob/develop/tests/Integration/JsonMapperTest.php#L463) on how to use it.

This has save lots of my time and effort where I was getting numeric keys in json.
